pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/deployment.go

```diff
@@ -51,17 +53,35 @@ var (
 	}
 )
 
+func getReplicas(cluster *kubermaticv1.Cluster) (*int32, error) {
+	if coredns := cluster.Spec.ComponentsOverride.CoreDNS; coredns != nil && coredns.Replicas != nil {
+		if network := cluster.Spec.ClusterNetwork; network.CoreDNSReplicas != nil && *network.CoreDNSReplicas != *coredns.Replicas {
+			return nil, errors.New("both the new spec.componentsOverride.coreDNS.replicas and deprecated spec.clusterNetwork.coreDNSReplicas fields are set")
+		}
```

Comment:
This should also be checked in `Cluster` validation so that invalid clusters get rejected.